staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go

```diff
@@ -28,9 +28,16 @@ type MessageCountMap map[string]int
 
 // Aggregate represents an object that contains multiple errors, but does not
 // necessarily have singular semantic meaning.
+// The aggregate can be used with `errors.Is()` to check for specific errors
+// and `errors.As()` to associate the first error of a specific type to a
+// target variable of the same type, if present.
+// It does not contain an `Unwrap` method, because the aggregated errors do
```

Comment:
I don't buy this argument, `Is` and `As` effectively define a search order; I think it would be logical if `Unwrap` returned the errors in the same order.

Reply:
The problem is that `Unwrap` assumes a 1 to 1 relationship of wrapping <-> wrapped error. If someone uses an `Unwrap` on an aggregate, they will get back one error and completely lose the rest. Can you describe a usecase where this is wanted?

The purpose of both `Is` and `As` here is to find out if a certain error among possibly more errors occurred. The usecase I have for that is retrying, I want to retry when a specific error happens and presumably as long as it happens, so I don't actually care which of the potentially more than one errors this matches on.

We could probably change the godoc to clarify that we don't provide any stability guarantee around the search order, WDYT?

Reply:
I am not saying that you need to implement Unwrap, just that there is a search order (which I think is correct!) and that would be fine for Unwrap to return errors in DFS order even though that loses the structure of what error is an ancestor/sibling of what. Unfortunately it does seem like the signature of Unwrap prevents e.g. doing a DFS and then returning the results one by one.

I do think this implementation of `Is`/`As` is potentially misleading. It could find a matching error that's not in a relevant branch, and also there's no way to find out if multiple errors match. You've really implemented e.g. `ContainsExact` and `ContainsType`. Would it be problematic to actually name these accordingly, and e.g. make them return a list of matching errors? Or is there some use that really requires the `errors.Is` call to work? If there is then I'm fine with this.

Reply:
Oh, and I think you *have* to guarantee the search order, otherwise code that worked could mysteriously stop working. Especially since you can only see the first matching error.

Reply:
> I do think this implementation of Is/As is potentially misleading.

Yeah, it may be misleading. The reason its named like that is to make it work with `errors.Is` which is the [official way](https://blog.golang.org/go1.13-errors) of finding out if an error is of a certain type and has the advantage that it continues working on errors that are being wrapped.

The intention behind `errors.As` is to make it work in conjunction with the [comparison functions in the `apierrors` package], which need to inspect some values, so just a boolean is not enough for them. Writing this down I am just realizing that this will not really help, because we just return the first error that matches the type, but may not match the matcher function, even if some other error in the `Aggregate` would. Hm, now I am inclined to to remove the `As` again because it feels like its too easy to use wrong. WDYT?

Reply:
SGTM, keep `Is`, remove or modify `As`.

Reply:
Could we could expand the `APIStatus` interface to implement `As` in a way that is aware of `ReasonForError`?  Just a thought.

Reply:
Actually never mind.  The implementation of `errors.As` is not super flexible and I think we would have to define custom container types for each reason.  Probably not worth it.

```go
	for err != nil {
		if reflectlite.TypeOf(err).AssignableTo(targetType) {
			val.Elem().Set(reflectlite.ValueOf(err))
			return true
		}
		if x, ok := err.(interface{ As(interface{}) bool }); ok && x.As(target) {
			return true
		}
		err = Unwrap(err)
	}
```

Reply:
s/is not super flexible/is super not flexible/

!